diff -urN TeXLive_svn_r25014/Build/source/texk/makejvf/Makefile.am uptex-1.00-20120115a/Build/source/texk/makejvf/Makefile.am
--- TeXLive_svn_r25014/Build/source/texk/makejvf/Makefile.am	2011-05-07 17:46:35 +0900
+++ uptex-1.00-20120115a/Build/source/texk/makejvf/Makefile.am	2011-12-24 06:23:45 +0900
@@ -11,7 +11,7 @@
 bin_PROGRAMS = makejvf
 
 makejvf_SOURCES = \
-	main.c  makejvf.h  tfmread.c  tool.c  write.c
+	main.c  makejvf.h  tfmread.c  tool.c  write.c  uniblock.c  uniblock.h
 
 $(makejvf_OBJECTS): $(KPATHSEA_DEPEND) $(PTEXENC_DEPEND)
 
diff -urN TeXLive_svn_r25014/Build/source/texk/makejvf/Makefile.in uptex-1.00-20120115a/Build/source/texk/makejvf/Makefile.in
--- TeXLive_svn_r25014/Build/source/texk/makejvf/Makefile.in	2011-12-23 17:21:02 +0900
+++ uptex-1.00-20120115a/Build/source/texk/makejvf/Makefile.in	2011-12-24 06:25:28 +0900
@@ -71,7 +71,7 @@
 am__installdirs = "$(DESTDIR)$(bindir)"
 PROGRAMS = $(bin_PROGRAMS)
 am_makejvf_OBJECTS = main.$(OBJEXT) tfmread.$(OBJEXT) tool.$(OBJEXT) \
-	write.$(OBJEXT)
+	write.$(OBJEXT) uniblock.$(OBJEXT)
 makejvf_OBJECTS = $(am_makejvf_OBJECTS)
 makejvf_LDADD = $(LDADD)
 am__DEPENDENCIES_1 =
@@ -234,7 +234,7 @@
 INCLUDES = $(PTEXENC_INCLUDES) $(KPATHSEA_INCLUDES)
 AM_CFLAGS = $(WARNING_CFLAGS)
 makejvf_SOURCES = \
-	main.c  makejvf.h  tfmread.c  tool.c  write.c
+	main.c  makejvf.h  tfmread.c  tool.c  write.c  uniblock.c  uniblock.h
 
 LDADD = $(PTEXENC_LIBS) $(KPATHSEA_LIBS)
 EXTRA_DIST = COPYRIGHT COPYRIGHT.jis Changes.txt README.txt \
@@ -354,6 +354,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/tfmread.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/tool.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/write.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/uniblock.Po@am__quote@
 
 .c.o:
 @am__fastdepCC_TRUE@	$(COMPILE) -MT $@ -MD -MP -MF $(DEPDIR)/$*.Tpo -c -o $@ $<
diff -urN TeXLive_svn_r25014/Build/source/texk/makejvf/main.c uptex-1.00-20120115a/Build/source/texk/makejvf/main.c
--- TeXLive_svn_r25014/Build/source/texk/makejvf/main.c	2011-05-07 17:46:35 +0900
+++ uptex-1.00-20120115a/Build/source/texk/makejvf/main.c	2011-12-24 09:48:12 +0900
@@ -5,14 +5,17 @@
 #include <kpathsea/config.h>
 #include <ptexenc/ptexenc.h>
 #include "makejvf.h"
+#include "uniblock.h"
 
 FILE *vfp,*afp=NULL;
-char *atfmname,*vtfmname,*afmname,*vfname,*kanatfm;
-int kanatume=-1,chotai=0,baseshift=0,minute=0;
+char *atfmname,*vtfmname,*afmname,*vfname,*kanatfm,*jistfm,*ucsqtfm;
+int kanatume=-1,chotai=0,baseshift=0,minute=0,useset3=0,hankana=0,fidzero=0;
+long ucs=0;
 
 int main(int argc, char ** argv)
 {
-	int i,j;
+	int i,j,ib;
+	long ch;
 
 	set_enc_string(NULL, "EUC");
 	if (argc < 3) {
@@ -66,6 +69,48 @@
 		case 'm':
 			minute=1;
 			break;
+		case 'u':
+			argv++;
+			if (!strcmp(*argv,"gb"))
+				ucs = ENTRY_G;
+			else if (!strcmp(*argv,"cns"))
+				ucs = ENTRY_C;
+			else if (!strcmp(*argv,"jisq"))
+				ucs = ENTRY_JQ;
+			else if (!strcmp(*argv,"jis"))
+				ucs = ENTRY_J;
+			else if (!strcmp(*argv,"ks"))
+				ucs = ENTRY_K;
+			else {
+				fprintf(stderr,"Charset is not set\n");
+				ucs = ENTRY_NO;
+			}
+			break;
+		case '3':
+			useset3=1;
+			break;
+		case 'J':
+			if ((*argv)[2]!='\0') {
+				jistfm = strdup(&(*argv)[2]);
+			}
+			else {
+				jistfm = strdup(*(++argv));
+			}
+			break;
+		case 'U':
+			if ((*argv)[2]!='\0') {
+				ucsqtfm = strdup(&(*argv)[2]);
+			}
+			else {
+				ucsqtfm = strdup(*(++argv));
+			}
+			break;
+		case 'H':
+			hankana=1;
+			break;
+		case 'i':
+			fidzero=1;
+			break;
 		default:
 			usage();
 			exit(0);
@@ -111,11 +156,35 @@
 		maketfm(kanatfm);
 	}
 
+	if (jistfm) {
+		if (!strcmp(&jistfm[strlen(jistfm)-4],".tfm")) {
+			jistfm[strlen(jistfm)-4] = '\0';
+		}
+		maketfm(jistfm);
+	}
+
+	if (ucsqtfm) {
+		if (!strcmp(&ucsqtfm[strlen(ucsqtfm)-4],".tfm")) {
+			ucsqtfm[strlen(ucsqtfm)-4] = '\0';
+		}
+		maketfm(ucsqtfm);
+	}
+
 	vfp = vfopen(vfname);
 
-	for (i=0;i<94;i++)
-		for (j=0;j<94;j++)
-			writevf((0x21+i)*256+(0x21+j),vfp);
+	if (ucs) {
+		ib=0;
+		for (i=0;i<(useset3+1);i++)
+			for (j=0;j<65536;j++) {
+				ch=i*2*65536+j;
+				if (search_cjk_entry(&ib,ch,ucs))
+					writevfu(ch,vfp);
+			}
+	} else {
+		for (i=0;i<94;i++)
+			for (j=0;j<94;j++)
+				writevf((0x21+i)*256+(0x21+j),vfp);
+	}
 
 	vfclose(vfp);
 
@@ -124,7 +193,7 @@
 
 void usage(void)
 {
-	fputs2("MAKEJVF ver.1.1a -- make Japanese VF file.\n", stderr);
+	fputs2("MAKEJVF ver.1.1a-u1.00 -- make Japanese VF file.\n", stderr);
 	fputs2("%% makejvf [<options>] <TFMfile> <PSfontTFM>\n", stderr);
 	fputs2("options:\n", stderr);
 	fputs2("-C           長体モード\n", stderr);
@@ -136,4 +205,12 @@
 	fputs2("-a <AFMfile> AFMファイル名（かな詰め時に使用）\n", stderr);
 	fputs2("-k <数値>    かな詰めマージン指定\n", stderr);
 	fputs2("             文字幅を1000として整数で指定。-aオプションと共に使用\n", stderr);
+	fputs2("-u <Charset> UCS mode\n", stderr);
+	fputs2("             <Charset> gb : GB,  cns : CNS,  ks : KS\n", stderr);
+	fputs2("                       jis : JIS,  jisq : JIS quote only\n", stderr);
+	fputs2("-J <TFMfile> JIS encoded PS font TFM name for quote, double quote with UCS mode\n", stderr);
+	fputs2("-U <TFMfile> UCS encoded PS font TFM name for quote, double quote with UCS mode\n", stderr);
+	fputs2("-3           use set3 (with UCS mode)\n", stderr);
+	fputs2("-H           use half-width katakana (with UCS mode)\n", stderr);
+	fputs2("-i           font ID from No.0\n", stderr);
 }
diff -urN TeXLive_svn_r25014/Build/source/texk/makejvf/makejvf.h uptex-1.00-20120115a/Build/source/texk/makejvf/makejvf.h
--- TeXLive_svn_r25014/Build/source/texk/makejvf/makejvf.h	2011-05-07 17:46:35 +0900
+++ uptex-1.00-20120115a/Build/source/texk/makejvf/makejvf.h	2011-12-24 06:14:53 +0900
@@ -1,9 +1,10 @@
 
 
-extern char *vtfmname,*kanatfm;
+extern char *vtfmname,*kanatfm,*jistfm,*ucsqtfm;
 extern int unit,zh,zw,jfm_id;
-extern int kanatume,chotai,baseshift,minute;
+extern int kanatume,chotai,baseshift,minute,hankana,fidzero;
 extern FILE *afp;
+extern long ucs;
 
 /* main.c */
 void usage(void);
@@ -27,5 +28,6 @@
 /* write.c */
 FILE *vfopen(char *name);
 void writevf(int code, FILE *fp);
+void writevfu(int code, FILE *fp);
 void vfclose(FILE *fp);
 void maketfm(char *name);
diff -urN TeXLive_svn_r25014/Build/source/texk/makejvf/uniblock.c uptex-1.00-20120115a/Build/source/texk/makejvf/uniblock.c
--- TeXLive_svn_r25014/Build/source/texk/makejvf/uniblock.c	1970-01-01 09:00:00 +0900
+++ uptex-1.00-20120115a/Build/source/texk/makejvf/uniblock.c	2011-12-23 18:43:09 +0900
@@ -0,0 +1,271 @@
+
+#ifdef DEBUG
+#include <stdio.h>
+#endif
+
+#include "uniblock.h"
+
+struct ublock {
+  long min, max, cjk;
+};
+
+/*
+ References:
+ [1] http://www.unicode.org/Public/UNIDATA/Blocks.txt
+        Date: 2010-06-04, 11:12:00 PDT [KW]
+ [2] CMap files
+     http://opensource.adobe.com/wiki/display/cmap/
+     cmapresources_cns1-6.tar.z     Adobe-CNS1-6 materials
+        cid2code.txt (Version 04/13/2010)
+     cmapresources_gb1-5.tar.z      Adobe-GB1-5 materials
+        cid2code.txt (Version 04/13/2010)
+     cmapresources_japan1-6.tar.z   Adobe-Japan1-6 materials
+        cid2code.txt (Version 06/24/2010)
+     cmapresources_korean1-2.tar.z  Adobe-Korea1-2 materials
+        cid2code.txt (Version 08/11/2009)
+*/
+
+static struct ublock ublock_data[] = {
+  {0x0000, 0x007F, ENTRY_GCJK}, /* Basic Latin */
+  {0x0080, 0x00FF, ENTRY_GCJK}, /* Latin-1 Supplement */
+  {0x0100, 0x017F, ENTRY_GCJK}, /* Latin Extended-A */
+  {0x0180, 0x024F, ENTRY_GCJ }, /* Latin Extended-B */
+  {0x0250, 0x02AF, ENTRY_GCJ }, /* IPA Extensions */
+  {0x02B0, 0x02FF, ENTRY_GCJK}, /* Spacing Modifier Letters */
+  {0x0300, 0x036F, ENTRY_CJ  }, /* Combining Diacritical Marks */
+  {0x0370, 0x03FF, ENTRY_GCJK}, /* Greek and Coptic */
+  {0x0400, 0x04FF, ENTRY_GCJK}, /* Cyrillic */
+  {0x0500, 0x052F, ENTRY_NO  }, /* Cyrillic Supplement */
+  {0x0530, 0x058F, ENTRY_NO  }, /* Armenian */
+  {0x0590, 0x05FF, ENTRY_NO  }, /* Hebrew */
+  {0x0600, 0x06FF, ENTRY_NO  }, /* Arabic */
+  {0x0700, 0x074F, ENTRY_NO  }, /* Syriac */
+  {0x0750, 0x077F, ENTRY_NO  }, /* Arabic Supplement */
+  {0x0780, 0x07BF, ENTRY_NO  }, /* Thaana */
+  {0x07C0, 0x07FF, ENTRY_NO  }, /* NKo */
+  {0x0800, 0x083F, ENTRY_NO  }, /* Samaritan */
+  {0x0840, 0x085F, ENTRY_NO  }, /* Mandaic */
+  {0x0900, 0x097F, ENTRY_NO  }, /* Devanagari */
+  {0x0980, 0x09FF, ENTRY_NO  }, /* Bengali */
+  {0x0A00, 0x0A7F, ENTRY_NO  }, /* Gurmukhi */
+  {0x0A80, 0x0AFF, ENTRY_NO  }, /* Gujarati */
+  {0x0B00, 0x0B7F, ENTRY_NO  }, /* Oriya */
+  {0x0B80, 0x0BFF, ENTRY_NO  }, /* Tamil */
+  {0x0C00, 0x0C7F, ENTRY_NO  }, /* Telugu */
+  {0x0C80, 0x0CFF, ENTRY_NO  }, /* Kannada */
+  {0x0D00, 0x0D7F, ENTRY_NO  }, /* Malayalam */
+  {0x0D80, 0x0DFF, ENTRY_NO  }, /* Sinhala */
+  {0x0E00, 0x0E7F, ENTRY_NO  }, /* Thai */
+  {0x0E80, 0x0EFF, ENTRY_NO  }, /* Lao */
+  {0x0F00, 0x0FFF, ENTRY_NO  }, /* Tibetan */
+  {0x1000, 0x109F, ENTRY_NO  }, /* Myanmar */
+  {0x10A0, 0x10FF, ENTRY_NO  }, /* Georgian */
+  {0x1100, 0x11FF, ENTRY_K   }, /* Hangul Jamo */
+  {0x1200, 0x137F, ENTRY_NO  }, /* Ethiopic */
+  {0x1380, 0x139F, ENTRY_NO  }, /* Ethiopic Supplement */
+  {0x13A0, 0x13FF, ENTRY_NO  }, /* Cherokee */
+  {0x1400, 0x167F, ENTRY_NO  }, /* Unified Canadian Aboriginal Syllabics */
+  {0x1680, 0x169F, ENTRY_NO  }, /* Ogham */
+  {0x16A0, 0x16FF, ENTRY_NO  }, /* Runic */
+  {0x1700, 0x171F, ENTRY_NO  }, /* Tagalog */
+  {0x1720, 0x173F, ENTRY_NO  }, /* Hanunoo */
+  {0x1740, 0x175F, ENTRY_NO  }, /* Buhid */
+  {0x1760, 0x177F, ENTRY_NO  }, /* Tagbanwa */
+  {0x1780, 0x17FF, ENTRY_NO  }, /* Khmer */
+  {0x1800, 0x18AF, ENTRY_NO  }, /* Mongolian */
+  {0x18B0, 0x18FF, ENTRY_NO  }, /* Unified Canadian Aboriginal Syllabics Extended */
+  {0x1900, 0x194F, ENTRY_NO  }, /* Limbu */
+  {0x1950, 0x197F, ENTRY_NO  }, /* Tai Le */
+  {0x1980, 0x19DF, ENTRY_NO  }, /* New Tai Lue */
+  {0x19E0, 0x19FF, ENTRY_NO  }, /* Khmer Symbols */
+  {0x1A00, 0x1A1F, ENTRY_NO  }, /* Buginese */
+  {0x1A20, 0x1AAF, ENTRY_NO  }, /* Tai Tham */
+  {0x1B00, 0x1B7F, ENTRY_NO  }, /* Balinese */
+  {0x1B80, 0x1BBF, ENTRY_NO  }, /* Sundanese */
+  {0x1BC0, 0x1BFF, ENTRY_NO  }, /* Batak */
+  {0x1C00, 0x1C4F, ENTRY_NO  }, /* Lepcha */
+  {0x1C50, 0x1C7F, ENTRY_NO  }, /* Ol Chiki */
+  {0x1CD0, 0x1CFF, ENTRY_NO  }, /* Vedic Extensions */
+  {0x1D00, 0x1D7F, ENTRY_NO  }, /* Phonetic Extensions */
+  {0x1D80, 0x1DBF, ENTRY_NO  }, /* Phonetic Extensions Supplement */
+  {0x1DC0, 0x1DFF, ENTRY_NO  }, /* Combining Diacritical Marks Supplement */
+  {0x1E00, 0x1EFF, ENTRY_GCJ }, /* Latin Extended Additional */
+  {0x1F00, 0x1FFF, ENTRY_J   }, /* Greek Extended */
+  {0x2000, 0x206F, ENTRY_GCJK}, /* General Punctuation */
+  {0x2070, 0x209F, ENTRY_JK  }, /* Superscripts and Subscripts */
+  {0x20A0, 0x20CF, ENTRY_GCJK}, /* Currency Symbols */
+  {0x20D0, 0x20FF, ENTRY_J   }, /* Combining Diacritical Marks for Symbols */
+  {0x2100, 0x214F, ENTRY_GCJK}, /* Letterlike Symbols */
+  {0x2150, 0x218F, ENTRY_GCJK}, /* Number Forms */
+  {0x2190, 0x21FF, ENTRY_GCJK}, /* Arrows */
+  {0x2200, 0x22FF, ENTRY_GCJK}, /* Mathematical Operators */
+  {0x2300, 0x23FF, ENTRY_GCJK}, /* Miscellaneous Technical */
+  {0x2400, 0x243F, ENTRY_CJ  }, /* Control Pictures */
+  {0x2440, 0x245F, ENTRY_NO  }, /* Optical Character Recognition */
+  {0x2460, 0x24FF, ENTRY_GCJK}, /* Enclosed Alphanumerics */
+  {0x2500, 0x257F, ENTRY_GCJK}, /* Box Drawing */
+  {0x2580, 0x259F, ENTRY_GCJK}, /* Block Elements */
+  {0x25A0, 0x25FF, ENTRY_GCJK}, /* Geometric Shapes */
+  {0x2600, 0x26FF, ENTRY_GCJK}, /* Miscellaneous Symbols */
+  {0x2700, 0x27BF, ENTRY_CJK }, /* Dingbats */
+  {0x27C0, 0x27EF, ENTRY_NO  }, /* Miscellaneous Mathematical Symbols-A */
+  {0x27F0, 0x27FF, ENTRY_NO  }, /* Supplemental Arrows-A */
+  {0x2800, 0x28FF, ENTRY_NO  }, /* Braille Patterns */
+  {0x2900, 0x297F, ENTRY_J   }, /* Supplemental Arrows-B */
+  {0x2980, 0x29FF, ENTRY_J   }, /* Miscellaneous Mathematical Symbols-B */
+  {0x2A00, 0x2AFF, ENTRY_NO  }, /* Supplemental Mathematical Operators */
+  {0x2B00, 0x2BFF, ENTRY_J   }, /* Miscellaneous Symbols and Arrows */
+  {0x2C00, 0x2C5F, ENTRY_NO  }, /* Glagolitic */
+  {0x2C60, 0x2C7F, ENTRY_NO  }, /* Latin Extended-C */
+  {0x2C80, 0x2CFF, ENTRY_NO  }, /* Coptic */
+  {0x2D00, 0x2D2F, ENTRY_NO  }, /* Georgian Supplement */
+  {0x2D30, 0x2D7F, ENTRY_NO  }, /* Tifinagh */
+  {0x2D80, 0x2DDF, ENTRY_NO  }, /* Ethiopic Extended */
+  {0x2DE0, 0x2DFF, ENTRY_NO  }, /* Cyrillic Extended-A */
+  {0x2E00, 0x2E7F, ENTRY_NO  }, /* Supplemental Punctuation */
+  {0x2E80, 0x2EFF, ENTRY_GCJ }, /* CJK Radicals Supplement */
+  {0x2F00, 0x2FDF, ENTRY_GCJK}, /* Kangxi Radicals */
+  {0x2FF0, 0x2FFF, ENTRY_G   }, /* Ideographic Description Characters */
+  {0x3000, 0x303F, ENTRY_GCJK}, /* CJK Symbols and Punctuation */
+  {0x3040, 0x309F, ENTRY_GCJK}, /* Hiragana */
+  {0x30A0, 0x30FF, ENTRY_GCJK}, /* Katakana */
+  {0x3100, 0x312F, ENTRY_GC  }, /* Bopomofo */
+  {0x3130, 0x318F, ENTRY_K   }, /* Hangul Compatibility Jamo */
+  {0x3190, 0x319F, ENTRY_J   }, /* Kanbun */
+  {0x31A0, 0x31BF, ENTRY_G   }, /* Bopomofo Extended */
+  {0x31C0, 0x31EF, ENTRY_C   }, /* CJK Strokes */
+  {0x31F0, 0x31FF, ENTRY_J   }, /* Katakana Phonetic Extensions */
+  {0x3200, 0x32FF, ENTRY_GCJK}, /* Enclosed CJK Letters and Months */
+  {0x3300, 0x33FF, ENTRY_GCJK}, /* CJK Compatibility */
+  {0x3400, 0x4DBF, ENTRY_GCJ }, /* CJK Unified Ideographs Extension A */
+  {0x4DC0, 0x4DFF, ENTRY_NO  }, /* Yijing Hexagram Symbols */
+  {0x4E00, 0x9FFF, ENTRY_GCJK}, /* CJK Unified Ideographs */
+  {0xA000, 0xA48F, ENTRY_G   }, /* Yi Syllables */
+  {0xA490, 0xA4CF, ENTRY_G   }, /* Yi Radicals */
+  {0xA4D0, 0xA4FF, ENTRY_NO  }, /* Lisu */
+  {0xA500, 0xA63F, ENTRY_NO  }, /* Vai */
+  {0xA640, 0xA69F, ENTRY_NO  }, /* Cyrillic Extended-B */
+  {0xA6A0, 0xA6FF, ENTRY_NO  }, /* Bamum */
+  {0xA700, 0xA71F, ENTRY_NO  }, /* Modifier Tone Letters */
+  {0xA720, 0xA7FF, ENTRY_NO  }, /* Latin Extended-D */
+  {0xA800, 0xA82F, ENTRY_NO  }, /* Syloti Nagri */
+  {0xA830, 0xA83F, ENTRY_NO  }, /* Common Indic Number Forms */
+  {0xA840, 0xA87F, ENTRY_NO  }, /* Phags-pa */
+  {0xA880, 0xA8DF, ENTRY_NO  }, /* Saurashtra */
+  {0xA8E0, 0xA8FF, ENTRY_NO  }, /* Devanagari Extended */
+  {0xA900, 0xA92F, ENTRY_NO  }, /* Kayah Li */
+  {0xA930, 0xA95F, ENTRY_NO  }, /* Rejang */
+  {0xA960, 0xA97F, ENTRY_NO  }, /* Hangul Jamo Extended-A */
+  {0xA980, 0xA9DF, ENTRY_NO  }, /* Javanese */
+  {0xAA00, 0xAA5F, ENTRY_NO  }, /* Cham */
+  {0xAA60, 0xAA7F, ENTRY_NO  }, /* Myanmar Extended-A */
+  {0xAA80, 0xAADF, ENTRY_NO  }, /* Tai Viet */
+  {0xAB00, 0xAB2F, ENTRY_NO  }, /* Ethiopic Extended-A */
+  {0xABC0, 0xABFF, ENTRY_NO  }, /* Meetei Mayek */
+  {0xAC00, 0xD7AF, ENTRY_K   }, /* Hangul Syllables */
+  {0xD7B0, 0xD7FF, ENTRY_NO  }, /* Hangul Jamo Extended-B */
+  {0xD800, 0xDB7F, ENTRY_NO  }, /* High Surrogates */
+  {0xDB80, 0xDBFF, ENTRY_NO  }, /* High Private Use Surrogates */
+  {0xDC00, 0xDFFF, ENTRY_NO  }, /* Low Surrogates */
+  {0xE000, 0xF8FF, ENTRY_GCJK}, /* Private Use Area */
+  {0xF900, 0xFAFF, ENTRY_GCJK}, /* CJK Compatibility Ideographs */
+  {0xFB00, 0xFB4F, ENTRY_J   }, /* Alphabetic Presentation Forms */
+  {0xFB50, 0xFDFF, ENTRY_NO  }, /* Arabic Presentation Forms-A */
+  {0xFE00, 0xFE0F, ENTRY_NO  }, /* Variation Selectors */
+  {0xFE10, 0xFE1F, ENTRY_GCJ }, /* Vertical Forms */
+  {0xFE20, 0xFE2F, ENTRY_NO  }, /* Combining Half Marks */
+  {0xFE30, 0xFE4F, ENTRY_GCJ }, /* CJK Compatibility Forms */
+  {0xFE50, 0xFE6F, ENTRY_GC  }, /* Small Form Variants */
+  {0xFE70, 0xFEFF, ENTRY_NO  }, /* Arabic Presentation Forms-B */
+  {0xFF00, 0xFFEF, ENTRY_GCJK}, /* Halfwidth and Fullwidth Forms */
+  {0xFFF0, 0xFFFF, ENTRY_NO  }, /* Specials */
+  {0x10000, 0x1007F, ENTRY_NO  }, /* Linear B Syllabary */
+  {0x10080, 0x100FF, ENTRY_NO  }, /* Linear B Ideograms */
+  {0x10100, 0x1013F, ENTRY_NO  }, /* Aegean Numbers */
+  {0x10140, 0x1018F, ENTRY_NO  }, /* Ancient Greek Numbers */
+  {0x10190, 0x101CF, ENTRY_NO  }, /* Ancient Symbols */
+  {0x101D0, 0x101FF, ENTRY_NO  }, /* Phaistos Disc */
+  {0x10280, 0x1029F, ENTRY_NO  }, /* Lycian */
+  {0x102A0, 0x102DF, ENTRY_NO  }, /* Carian */
+  {0x10300, 0x1032F, ENTRY_NO  }, /* Old Italic */
+  {0x10330, 0x1034F, ENTRY_NO  }, /* Gothic */
+  {0x10380, 0x1039F, ENTRY_NO  }, /* Ugaritic */
+  {0x103A0, 0x103DF, ENTRY_NO  }, /* Old Persian */
+  {0x10400, 0x1044F, ENTRY_NO  }, /* Deseret */
+  {0x10450, 0x1047F, ENTRY_NO  }, /* Shavian */
+  {0x10480, 0x104AF, ENTRY_NO  }, /* Osmanya */
+  {0x10800, 0x1083F, ENTRY_NO  }, /* Cypriot Syllabary */
+  {0x10840, 0x1085F, ENTRY_NO  }, /* Imperial Aramaic */
+  {0x10900, 0x1091F, ENTRY_NO  }, /* Phoenician */
+  {0x10920, 0x1093F, ENTRY_NO  }, /* Lydian */
+  {0x10A00, 0x10A5F, ENTRY_NO  }, /* Kharoshthi */
+  {0x10A60, 0x10A7F, ENTRY_NO  }, /* Old South Arabian */
+  {0x10B00, 0x10B3F, ENTRY_NO  }, /* Avestan */
+  {0x10B40, 0x10B5F, ENTRY_NO  }, /* Inscriptional Parthian */
+  {0x10B60, 0x10B7F, ENTRY_NO  }, /* Inscriptional Pahlavi */
+  {0x10C00, 0x10C4F, ENTRY_NO  }, /* Old Turkic */
+  {0x10E60, 0x10E7F, ENTRY_NO  }, /* Rumi Numeral Symbols */
+  {0x11000, 0x1107F, ENTRY_NO  }, /* Brahmi */
+  {0x11080, 0x110CF, ENTRY_NO  }, /* Kaithi */
+  {0x12000, 0x123FF, ENTRY_NO  }, /* Cuneiform */
+  {0x12400, 0x1247F, ENTRY_NO  }, /* Cuneiform Numbers and Punctuation */
+  {0x13000, 0x1342F, ENTRY_NO  }, /* Egyptian Hieroglyphs */
+  {0x16800, 0x16A3F, ENTRY_NO  }, /* Bamum Supplement */
+  {0x1B000, 0x1B0FF, ENTRY_NO  }, /* Kana Supplement */
+  {0x1D000, 0x1D0FF, ENTRY_NO  }, /* Byzantine Musical Symbols */
+  {0x1D100, 0x1D1FF, ENTRY_NO  }, /* Musical Symbols */
+  {0x1D200, 0x1D24F, ENTRY_NO  }, /* Ancient Greek Musical Notation */
+  {0x1D300, 0x1D35F, ENTRY_NO  }, /* Tai Xuan Jing Symbols */
+  {0x1D360, 0x1D37F, ENTRY_NO  }, /* Counting Rod Numerals */
+  {0x1D400, 0x1D7FF, ENTRY_NO  }, /* Mathematical Alphanumeric Symbols */
+  {0x1F000, 0x1F02F, ENTRY_NO  }, /* Mahjong Tiles */
+  {0x1F030, 0x1F09F, ENTRY_NO  }, /* Domino Tiles */
+  {0x1F0A0, 0x1F0FF, ENTRY_NO  }, /* Playing Cards */
+  {0x1F100, 0x1F1FF, ENTRY_NO  }, /* Enclosed Alphanumeric Supplement */
+  {0x1F200, 0x1F2FF, ENTRY_NO  }, /* Enclosed Ideographic Supplement */
+  {0x1F300, 0x1F5FF, ENTRY_NO  }, /* Miscellaneous Symbols And Pictographs */
+  {0x1F600, 0x1F64F, ENTRY_NO  }, /* Emoticons */
+  {0x1F680, 0x1F6FF, ENTRY_NO  }, /* Transport And Map Symbols */
+  {0x1F700, 0x1F77F, ENTRY_NO  }, /* Alchemical Symbols */
+  {0x20000, 0x2A6DF, ENTRY_GCJ }, /* CJK Unified Ideographs Extension B */
+  {0x2A700, 0x2B73F, ENTRY_CJ  }, /* CJK Unified Ideographs Extension C */
+  {0x2B740, 0x2B81F, ENTRY_J   }, /* CJK Unified Ideographs Extension D */
+  {0x2F800, 0x2FA1F, ENTRY_CJ  }, /* CJK Compatibility Ideographs Supplement */
+  {0xE0000, 0xE007F, ENTRY_NO  }, /* Tags */
+  {0xE0100, 0xE01EF, ENTRY_NO  }, /* Variation Selectors Supplement */
+  {0xF0000, 0xFFFFF, ENTRY_NO  }, /* Supplementary Private Use Area-A */
+  {0x100000, 0x10FFFF, ENTRY_NO  }, /* Supplementary Private Use Area-B */
+  {0x200000, 0xFFFFFF, ENTRY_NO  }, /* illegal */
+};
+
+
+int search_cjk_entry(int *ib, long ch, long cjk) {
+  if (cjk==ENTRY_NO) return 1;
+  if (cjk==ENTRY_JQ) return
+	(ch==U_OPEN_SQUOTE || ch==U_CLOSE_SQUOTE
+	 || ch==U_OPEN_DQUOTE || ch==U_CLOSE_DQUOTE);
+  while(ublock_data[*ib].max<ch) (*ib)++;
+  if (ublock_data[*ib].min<=ch && ch<=ublock_data[*ib].max) {
+    return ublock_data[*ib].cjk & cjk;
+  } else {
+    return 0;
+  }
+}
+
+
+#ifdef DEBUG
+int main() {
+  long ch;
+  int ib;
+
+  ib=0;
+  for (ch=0x0;ch<0x10000;ch++) {
+    printf(" %05x %2d %2d %2d %2d %2d\n", ch, ib,
+	   search_cjk_entry(&ib,ch,ENTRY_G),
+	   search_cjk_entry(&ib,ch,ENTRY_C),
+	   search_cjk_entry(&ib,ch,ENTRY_J),
+	   search_cjk_entry(&ib,ch,ENTRY_K));
+  }
+
+}
+#endif
diff -urN TeXLive_svn_r25014/Build/source/texk/makejvf/uniblock.h uptex-1.00-20120115a/Build/source/texk/makejvf/uniblock.h
--- TeXLive_svn_r25014/Build/source/texk/makejvf/uniblock.h	1970-01-01 09:00:00 +0900
+++ uptex-1.00-20120115a/Build/source/texk/makejvf/uniblock.h	2011-05-06 00:29:33 +0900
@@ -0,0 +1,27 @@
+
+#define ENTRY_NO    0x01
+/* for JIS, quote only */
+/* for GB, CNS, JIS, KS */
+#define ENTRY_JQ    0x1000
+#define ENTRY_G     0x02
+#define ENTRY_C     0x04
+#define ENTRY_J     0x08
+#define ENTRY_K     0x10
+#define ENTRY_GC    ENTRY_G|ENTRY_C
+#define ENTRY_GJ    ENTRY_G        |ENTRY_J
+#define ENTRY_GK    ENTRY_G                |ENTRY_K
+#define ENTRY_CJ            ENTRY_C|ENTRY_J
+#define ENTRY_CK            ENTRY_C        |ENTRY_K
+#define ENTRY_JK                    ENTRY_J|ENTRY_K
+#define ENTRY_GCJ   ENTRY_G|ENTRY_C|ENTRY_J
+#define ENTRY_GCK   ENTRY_G|ENTRY_C        |ENTRY_K
+#define ENTRY_GJK   ENTRY_G        |ENTRY_J|ENTRY_K
+#define ENTRY_CJK           ENTRY_C|ENTRY_J|ENTRY_K
+#define ENTRY_GCJK  ENTRY_G|ENTRY_C|ENTRY_J|ENTRY_K
+
+extern int search_cjk_entry(int *ib, long ch, long cjk);
+
+#define U_OPEN_SQUOTE   0x2018
+#define U_CLOSE_SQUOTE  0x2019
+#define U_OPEN_DQUOTE   0x201C
+#define U_CLOSE_DQUOTE  0x201D
diff -urN TeXLive_svn_r25014/Build/source/texk/makejvf/write.c uptex-1.00-20120115a/Build/source/texk/makejvf/write.c
--- TeXLive_svn_r25014/Build/source/texk/makejvf/write.c	2011-05-07 17:46:35 +0900
+++ uptex-1.00-20120115a/Build/source/texk/makejvf/write.c	2011-12-24 06:28:29 +0900
@@ -3,11 +3,14 @@
 
 #include <kpathsea/config.h>
 #include "makejvf.h"
+#include "uniblock.h"
 
 FILE *vfopen(char *name)
 {
 	FILE *fp;
+	int fidshift=0;
 
+	if (fidzero) fidshift=-1;
 	fp = fopen(name,"w");
 	if (fp == NULL) {
 		fprintf(stderr,"I cannot create VF file, %s.",name);
@@ -21,7 +24,7 @@
 	fputnum(10*(1<<20),4,fp); /* TFM design size */
 
 	fputc(243,fp); /* fnt_def1 */
-	fputc(1,fp); /* Font ID */
+	fputc(1+fidshift,fp); /* Font ID */
 	fputnum(0,4,fp); /* TFM check sum */
 	if (chotai)
 		fputnum(zh,4,fp); /* font design size (scaled) */
@@ -34,7 +37,7 @@
 
 	if (kanatfm) {
 		fputc(243,fp); /* fnt_def1 */
-		fputc(2,fp); /* Font ID */
+		fputc(2+fidshift,fp); /* Font ID */
 		fputnum(0,4,fp); /* TFM check sum */
 		if (chotai)
 			fputnum(zh,4,fp); /* font design size (scaled) */
@@ -46,6 +49,33 @@
 		fputstr(kanatfm,strlen(kanatfm),fp); /* directory + fontname */
 	}
 
+	if (ucsqtfm) {
+		fputc(243,fp); /* fnt_def1 */
+		fputc(3+fidshift,fp); /* Font ID */
+		fputnum(0,4,fp); /* TFM check sum */
+		if (chotai)
+			fputnum(zh,4,fp); /* font design size (scaled) */
+		else
+			fputnum(zw,4,fp); /* font design size (scaled) */
+		fputnum(10*(1<<20),4,fp); /* font design size */
+		fputc(0,fp); /* directory length */
+		fputc(strlen(ucsqtfm),fp); /* fontname length */
+		fputstr(ucsqtfm,strlen(ucsqtfm),fp); /* directory + fontname */
+	}
+	else if (jistfm) {
+		fputc(243,fp); /* fnt_def1 */
+		fputc(3+fidshift,fp); /* Font ID */
+		fputnum(0,4,fp); /* TFM check sum */
+		if (chotai)
+			fputnum(zh,4,fp); /* font design size (scaled) */
+		else
+			fputnum(zw,4,fp); /* font design size (scaled) */
+		fputnum(10*(1<<20),4,fp); /* font design size */
+		fputc(0,fp); /* directory length */
+		fputc(strlen(jistfm),fp); /* fontname length */
+		fputstr(jistfm,strlen(jistfm),fp); /* directory + fontname */
+	}
+
 	return fp;
 }
 
@@ -53,6 +83,9 @@
 {
 	int cc,cc2,cc3,cc4,w,skip=0,skip2=0,height=1000;
 	char buf[256],buf2[256];
+	int fidshift=0;
+
+	if (fidzero) fidshift=-1;
 
 	w = jfmread(code);
 
@@ -109,7 +142,7 @@
 				fputc(157+numcount(skip2)-1,fp); /* DOWN */
 				fputnum2(skip2,fp);
 			}
-			if (kanatfm) fputc(173,fp); /* FONT_NUM_2 */
+			if (kanatfm) fputc(173+fidshift,fp); /* FONT_NUM_2 */
 			fputc(129,fp); /* SET2 */
 			if (code == 0x2146)
 				fputnum(0x216c,2,fp); /* char code */
@@ -189,7 +222,7 @@
 			fputc(239,fp); /* XXX1 */
 			fputc(88,fp);
 			fputs("ps: gsave currentpoint currentpoint translate 180 neg rotate neg exch neg exch translate",fp);
-			if (kanatfm) fputc(173,fp); /* FONT_NUM_2 */
+			if (kanatfm) fputc(173+fidshift,fp); /* FONT_NUM_2 */
 			fputc(129,fp); /* SET2 */
 			if (code == 0x2147)
 				fputnum(0x216c,2,fp); /* char code */
@@ -328,14 +361,444 @@
 	}
 	if (kanatfm) {
 		if (code <= 0x2576)
-			fputc(173,fp); /* FONT_NUM_2 */
+			fputc(173+fidshift,fp); /* FONT_NUM_2 */
 		else
-			fputc(172,fp); /* FONT_NUM_1 */
+			fputc(172+fidshift,fp); /* FONT_NUM_1 */
 	}
 	fputc(129,fp); /* SET2 */
 	fputnum(code,2,fp); /* char code */
 }
 
+void writevfu(int code, FILE *fp)
+{
+	int cc,cc2,cc3,cc4,w,skip=0,skip2=0,height=1000;
+	char buf[256],buf2[256];
+	int fidshift=0;
+
+	if (fidzero) fidshift=-1;
+
+	w = jfmread(code);
+
+	fputc(242,fp); /* long_char */
+
+	skip2=baseshift;
+	switch (code) {
+	case 0x2018: /* ‘ */
+	case 0x201c: /* “ */
+		if (jfm_id == 9 && minute) { /* 縦書き時はミニュートへ変換 */
+			if (afp) {
+				if (code == 0x2018)
+					sprintf(buf2,"CH <2032>");
+				else
+					sprintf(buf2,"CH <2033>");
+				rewind(afp);
+				while (fgets(buf,255,afp)!=NULL) {
+					if (jfm_id==9 && !strncmp(buf,"FontBBox ",9)) {
+						sscanf(&buf[9],"%d %d %d %d",&cc,&cc2,&cc3,&cc4);
+						height=cc4;
+					}
+					if (!strncmp(buf,buf2,strlen(buf2))) {
+						sscanf(&buf[14],"%d %d %d %d",&cc,&cc2,&cc3,&cc4);
+						skip=(int)(w+((double)(cc2-height)/1000.0-0.05)*zw);
+						break;
+					}
+				}
+			}
+			else
+				skip=(int)((0.1)*zw);
+			if (code == 0x2018) {
+				skip2+=-(int)((0.65)*zh);
+			}
+			else {
+				skip2+=-(int)((0.6)*zh);
+			}
+
+			if (kanatfm)
+				cc=4;
+			else
+				cc=3;
+			if (skip)
+				cc+=numcount(skip)+1;
+			if (skip2)
+				cc+=numcount(skip2)+1;
+			fputnum(cc,4,fp);
+			fputnum(code,4,fp); /* char code */
+			fputnum(w,4,fp); /* char width */
+			if (skip) {
+				fputc(143+numcount(skip)-1,fp); /* RIGHT */
+				fputnum2(skip,fp);
+			}
+			if (skip2) {
+				fputc(157+numcount(skip2)-1,fp); /* DOWN */
+				fputnum2(skip2,fp);
+			}
+			if (kanatfm) fputc(173+fidshift,fp); /* FONT_NUM_2 */
+			if (code == 0x2018)
+				fputnum(0x2032,2,fp); /* char code */
+			else
+				fputnum(0x2033,2,fp); /* char code */
+			return;
+		}
+		else if (ucsqtfm) { /* UniJIS-UCS2-H系へ変換 */
+			cc=4;
+			skip = -(zw-w);
+			if (skip)
+				cc+=numcount(skip)+1;
+			if (skip2)
+				cc+=numcount(skip2)+1;
+			fputnum(cc,4,fp);
+			fputnum(code,4,fp); /* char code */
+			fputnum(w,4,fp); /* char width */
+			if (skip) {
+				fputc(143+numcount(skip)-1,fp); /* RIGHT */
+				fputnum2(skip,fp);
+			}
+			if (skip2) {
+				fputc(157+numcount(skip2)-1,fp); /* DOWN */
+				fputnum2(skip2,fp);
+			}
+			fputc(174+fidshift,fp); /* FONT_NUM_3 */
+			fputc(129,fp); /* SET2 */
+			fputnum(code,2,fp); /* char code */
+			return;
+		}
+		else if (jfm_id == 11 && jistfm) { /* 横書き時はJIS系へ変換 */
+		  /* UCS U+2018 → JIS 0x2146, UCS U+201C → JIS 0x2148 */
+			cc=4;
+			skip = -(zw-w);
+			if (skip)
+				cc+=numcount(skip)+1;
+			if (skip2)
+				cc+=numcount(skip2)+1;
+			fputnum(cc,4,fp);
+			fputnum(code,4,fp); /* char code */
+			fputnum(w,4,fp); /* char width */
+			if (skip) {
+				fputc(143+numcount(skip)-1,fp); /* RIGHT */
+				fputnum2(skip,fp);
+			}
+			if (skip2) {
+				fputc(157+numcount(skip2)-1,fp); /* DOWN */
+				fputnum2(skip2,fp);
+			}
+			fputc(174+fidshift,fp); /* FONT_NUM_3 */
+			fputc(129,fp); /* SET2 */
+			if (code == 0x2018)
+				fputnum(0x2146,2,fp); /* char code */
+			else
+				fputnum(0x2148,2,fp); /* char code */
+			return;
+		}
+	case 0xFF08: /* （ */
+	case 0x3014: /* 〔 */
+	case 0xFF3B: /* ［ */
+	case 0xFF5B: /* ｛ */
+	case 0x3008: /* 〈 */
+	case 0x300A: /* 《 */
+	case 0x300C: /* 「 */
+	case 0x300E: /* 『 */
+	case 0x3010: /* 【 */
+	case 0xFF5F: /* JIS X 0213  1-02-54 始め二重バーレーン */
+	case 0x3018: /* JIS X 0213  1-02-56 始め二重亀甲括弧 */
+	case 0x3016: /* JIS X 0213  1-02-58 始めすみ付き括弧(白) */
+	case 0x301D: /* JIS X 0213  1-13-64 始めダブルミニュート */
+		if (ucs != ENTRY_JQ)
+			skip = -(zw-w);
+		if (kanatfm)
+			cc=4;
+		else
+			cc=3;
+		if (skip)
+			cc+=numcount(skip)+1;
+		if (skip2)
+			cc+=numcount(skip2)+1;
+		fputnum(cc,4,fp);
+		break;
+	case 0x2019: /* ’ */
+	case 0x201d: /* ” */
+		if (jfm_id == 9 && minute) { /* 縦書き時はミニュートへ変換 */
+			if (afp) {
+				if (code == 0x2019)
+					sprintf(buf2,"CH <216C>");
+				else
+					sprintf(buf2,"CH <216D>");
+				rewind(afp);
+				while (fgets(buf,255,afp)!=NULL) {
+					if (jfm_id==9 && !strncmp(buf,"FontBBox ",9)) {
+						sscanf(&buf[9],"%d %d %d %d",&cc,&cc2,&cc3,&cc4);
+						height=cc4;
+					}
+					if (!strncmp(buf,buf2,strlen(buf2))) {
+						sscanf(&buf[14],"%d %d %d %d",&cc,&cc2,&cc3,&cc4);
+						skip=(int)(((double)(height-cc2)/1000.0+0.05)*zw);
+						break;
+					}
+				}
+			}
+			else
+				skip=(int)((0.4)*zw);
+			if (code == 0x2019) {
+				skip2+=(int)((0.65)*zh);
+			}
+			else {
+				skip2+=(int)((0.6)*zh);
+			}
+
+			if (kanatfm)
+				cc=4;
+			else
+				cc=3;
+			if (skip)
+				cc+=numcount(skip)+1;
+			if (skip2)
+				cc+=numcount(skip2)+1;
+			fputnum(cc+2+88+2+32,4,fp);
+			fputnum(code,4,fp); /* char code */
+			fputnum(w,4,fp); /* char width */
+			if (skip) {
+				fputc(143+numcount(skip)-1,fp); /* RIGHT */
+				fputnum2(skip,fp);
+			}
+			if (skip2) {
+				fputc(157+numcount(skip2)-1,fp); /* DOWN */
+				fputnum2(skip2,fp);
+			}
+			fputc(239,fp); /* XXX1 */
+			fputc(88,fp);
+			fputs("ps: gsave currentpoint currentpoint translate 180 neg rotate neg exch neg exch translate",fp);
+			if (kanatfm) fputc(173+fidshift,fp); /* FONT_NUM_2 */
+			fputc(129,fp); /* SET2 */
+			if (code == 0x2019)
+				fputnum(0x2032,2,fp); /* char code */
+			else
+				fputnum(0x2033,2,fp); /* char code */
+			fputc(239,fp); /* XXX1 */
+			fputc(32,fp);
+			fputs("ps: currentpoint grestore moveto",fp);
+			return;
+		}
+		else if (ucsqtfm) { /* UniJIS-UCS2-H系へ変換 */
+			cc=4;
+			if (skip2)
+				cc+=numcount(skip2)+1;
+			fputnum(cc,4,fp);
+			fputnum(code,4,fp); /* char code */
+			fputnum(w,4,fp); /* char width */
+			if (skip2) {
+				fputc(157+numcount(skip2)-1,fp); /* DOWN */
+				fputnum2(skip2,fp);
+			}
+			fputc(174+fidshift,fp); /* FONT_NUM_3 */
+			fputc(129,fp); /* SET2 */
+			fputnum(code,2,fp); /* char code */
+			return;
+		}
+		else if (jfm_id == 11 && jistfm) { /* 横書き時はJIS系へ変換 */
+		  /* UCS U+2019 → JIS 0x2147, UCS U+201D → JIS 0x2149 */
+			cc=4;
+			if (skip2)
+				cc+=numcount(skip2)+1;
+			fputnum(cc,4,fp);
+			fputnum(code,4,fp); /* char code */
+			fputnum(w,4,fp); /* char width */
+			if (skip2) {
+				fputc(157+numcount(skip2)-1,fp); /* DOWN */
+				fputnum2(skip2,fp);
+			}
+			fputc(174+fidshift,fp); /* FONT_NUM_3 */
+			fputc(129,fp); /* SET2 */
+			if (code == 0x2019)
+				fputnum(0x2147,2,fp); /* char code */
+			else
+				fputnum(0x2149,2,fp); /* char code */
+			return;
+		}
+	case 0x3000: /* spc */
+	case 0x3001: /* 、 */
+	case 0x3002: /* 。 */
+	case 0xFF0C: /* ， */
+	case 0xFF0E: /* ． */
+	case 0x309B: /* ゛ */
+	case 0x309C: /* ゜ */
+	case 0xFF09: /* ） */
+	case 0x3015: /* 〕 */
+	case 0xFF3D: /* ］ */
+	case 0xFF5D: /* ｝ */
+	case 0x3009: /* 〉 */
+	case 0x300B: /* 》 */
+	case 0x300D: /* 」 */
+	case 0x300F: /* 』 */
+	case 0x3011: /* 】 */
+	case 0xFF60: /* JIS X 0213  1-02-55 終わり二重バーレーン */
+	case 0x3019: /* JIS X 0213  1-02-57 終わり二重亀甲括弧 */
+	case 0x3017: /* JIS X 0213  1-02-59 終わりすみ付き括弧(白) */
+	case 0x301F: /* JIS X 0213  1-13-65 終わりダブルミニュート */
+	case 0x00B0: /* ° */
+	case 0x2032: /* ′ */
+	case 0x2033: /* ″ */
+		if (kanatfm)
+			cc=4;
+		else
+			cc=3;
+		if (skip2)
+			cc+=numcount(skip2)+1;
+		fputnum(cc,4,fp);
+		break;
+	case 0xFF61: case 0xFF62: case 0xFF63: case 0xFF64: case 0xFF65: case 0xFF66: case 0xFF67:
+	case 0xFF68: case 0xFF69: case 0xFF6A: case 0xFF6B: case 0xFF6C: case 0xFF6D: case 0xFF6E: case 0xFF6F:
+	case 0xFF70: case 0xFF71: case 0xFF72: case 0xFF73: case 0xFF74: case 0xFF75: case 0xFF76: case 0xFF77:
+	case 0xFF78: case 0xFF79: case 0xFF7A: case 0xFF7B: case 0xFF7C: case 0xFF7D: case 0xFF7E: case 0xFF7F:
+	case 0xFF80: case 0xFF81: case 0xFF82: case 0xFF83: case 0xFF84: case 0xFF85: case 0xFF86: case 0xFF87:
+	case 0xFF88: case 0xFF89: case 0xFF8A: case 0xFF8B: case 0xFF8C: case 0xFF8D: case 0xFF8E: case 0xFF8F:
+	case 0xFF90: case 0xFF91: case 0xFF92: case 0xFF93: case 0xFF94: case 0xFF95: case 0xFF96: case 0xFF97:
+	case 0xFF98: case 0xFF99: case 0xFF9A: case 0xFF9B: case 0xFF9C: case 0xFF9D: case 0xFF9E: case 0xFF9F:
+		if (jfm_id == 11 && hankana) { /* 半角片仮名、横書き時 */
+			cc=3;
+			if (skip2)
+				cc+=numcount(skip2)+1;
+			fputnum(cc,4,fp);
+			fputnum(code,4,fp); /* char code */
+			fputnum(w,4,fp); /* char width */
+			if (skip2) {
+				fputc(157+numcount(skip2)-1,fp); /* DOWN */
+				fputnum2(skip2,fp);
+			}
+			fputc(129,fp); /* SET2 */
+			fputnum(code,2,fp); /* char code */
+			return;
+		}
+	default:
+		if (w != zw) {
+			if (((code >= 0x3041 && code <= 0x30F6) || code == 0x30FC ) && kanatume>=0) {
+				sprintf(buf2,"CH <%X>",code);
+				rewind(afp);
+				while (fgets(buf,255,afp)!=NULL) {
+					if (jfm_id==9 && !strncmp(buf,"FontBBox ",9)) {
+						sscanf(&buf[9],"%d %d %d %d",&cc,&cc2,&cc3,&cc4);
+						height=cc4;
+					}
+					if (!strncmp(buf,buf2,strlen(buf2))) {
+						sscanf(&buf[14],"%d %d %d %d",&cc,&cc2,&cc3,&cc4);
+						if (jfm_id==9) {
+							switch (code) {
+							case 0x3041:
+							case 0x3043:
+							case 0x3045:
+							case 0x3047:
+							case 0x3049:
+							case 0x3063:
+							case 0x3083:
+							case 0x3085:
+							case 0x3087:
+							case 0x308E:
+							case 0x30A1:
+							case 0x30A3:
+							case 0x30A5:
+							case 0x30A7:
+							case 0x30A9:
+							case 0x30C3:
+							case 0x30E3:
+							case 0x30E5:
+							case 0x30E7:
+							case 0x30EE:
+							case 0x30F5:
+							case 0x30F6:
+							case 0x3095: /* JIS X 0213  1-04-85 小書き平仮名か */
+							case 0x3096: /* JIS X 0213  1-04-86 小書き平仮名け */
+							case 0x31F0: /* JIS X 0213  1-06-78 小書き片仮名ク */
+							case 0x31F1: /* JIS X 0213  1-06-79 小書き片仮名シ */
+							case 0x31F2: /* JIS X 0213  1-06-80 小書き片仮名ス */
+							case 0x31F3: /* JIS X 0213  1-06-81 小書き片仮名ト */
+							case 0x31F4: /* JIS X 0213  1-06-82 小書き片仮名ヌ */
+							case 0x31F5: /* JIS X 0213  1-06-83 小書き片仮名ハ */
+							case 0x31F6: /* JIS X 0213  1-06-84 小書き片仮名ヒ */
+							case 0x31F7: /* JIS X 0213  1-06-85 小書き片仮名フ */
+							case 0x31F8: /* JIS X 0213  1-06-86 小書き片仮名ヘ */
+							case 0x31F9: /* JIS X 0213  1-06-87 小書き片仮名ホ */
+							case 0x31FA: /* JIS X 0213  1-06-89 小書き片仮名ム */
+							case 0x31FB: /* JIS X 0213  1-06-90 小書き片仮名ラ */
+							case 0x31FC: /* JIS X 0213  1-06-91 小書き片仮名リ */
+							case 0x31FD: /* JIS X 0213  1-06-92 小書き片仮名ル */
+							case 0x31FE: /* JIS X 0213  1-06-93 小書き片仮名レ */
+							case 0x31FF: /* JIS X 0213  1-06-94 小書き片仮名ロ */
+								skip=-(int)(((double)(1000-(cc4-cc2)-kanatume*2)/2/1000.0)*zw);
+								break;
+							case 0x30FC:
+								skip=-(int)((double)(cc-kanatume)/1000.0*zw);
+								break;
+							default:
+								skip=-(int)(((double)(height-cc4-kanatume)/1000.0)*zw);
+								break;
+							}
+						}
+						else {
+							skip=-(int)(((double)(cc-kanatume)/1000.0)*zw);
+						}
+						if (kanatfm)
+							cc=4;
+						else
+							cc=3;
+						if (skip)
+							cc+=numcount(skip)+1;
+						if (skip2)
+							cc+=numcount(skip2)+1;
+						fputnum(cc,4,fp);
+						break;
+					}
+				}
+			}
+			else {
+				skip = -(zw-w)/2;
+				if (kanatfm)
+					cc=4;
+				else
+					cc=3;
+				if (skip)
+					cc+=numcount(skip)+1;
+				if (skip2)
+					cc+=numcount(skip2)+1;
+				fputnum(cc,4,fp);
+			}
+		}
+		else {
+			if (kanatfm || code>=0x10000)
+				cc=4;
+			else
+				cc=3;
+			if (skip)
+				cc+=numcount(skip)+1;
+			if (skip2)
+				cc+=numcount(skip2)+1;
+			fputnum(cc,4,fp);
+		}
+		break;
+	}
+
+	fputnum(code,4,fp); /* char code */
+	fputnum(w,4,fp); /* char width */
+	if (skip) {
+		fputc(143+numcount(skip)-1,fp); /* RIGHT */
+		fputnum2(skip,fp);
+	}
+	if (skip2) {
+		fputc(157+numcount(skip2)-1,fp); /* DOWN */
+		fputnum2(skip2,fp);
+	}
+	if (kanatfm) {
+		if (code <= 0x30F6)
+			fputc(173+fidshift,fp); /* FONT_NUM_2 */
+		else
+			fputc(172+fidshift,fp); /* FONT_NUM_1 */
+	}
+	if (code>=0x10000) {
+		fputc(130,fp); /* SET3 */
+		fputnum(code,3,fp); /* char code */
+	} else {
+		fputc(129,fp); /* SET2 */
+		fputnum(code,2,fp); /* char code */
+	}
+}
+
 void vfclose(FILE *fp)
 {
 	int i,cc;
diff -urN TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/ChangeLog uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/ChangeLog
--- TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/ChangeLog	2012-01-09 11:14:49 +0900
+++ uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/ChangeLog	2012-01-14 13:47:11 +0900
@@ -1,3 +1,12 @@
+2012-01-15  Takuji Tanaka  <KXD02663@nifty.ne.jp>
+
+	* uptex-m.ch, uppltotf.ch uptftopl.ch, updvitype.ch, upbibtex.ch:
+	Update upTeX version number.
+
+	* uptex-m.ch, kanji.c:
+	Update reference of Unicode blocks to
+	"Blocks-6.0.0.txt Date: 2010-06-04, 11:12:00 PDT [KW]".
+
 2012-01-07  Akira Kakuto  <kakuto@fuk.kindai.ac.jp>
 
 	* kanji.h: revert wrt init_kanji (from TANAKA Takuji-san).
diff -urN TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/kanji.c uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/kanji.c
--- TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/kanji.c	2011-12-22 23:17:34 +0900
+++ uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/kanji.c	2012-01-07 10:32:27 +0900
@@ -15,8 +15,8 @@
 int sjisterminal;
 #endif
 
-/*  TOKEN */
-boolean check_kanji (integer c)
+/* TOKEN */
+boolean check_kanji(integer c)
 {
     if (c > CS_TOKEN_FLAG) return false;
     else if (!(XXHi(c)>=KCAT_KANJI && XXHi(c)<=KCAT_HANGUL)) return false;
@@ -60,6 +60,8 @@
 }
 
 /* Ref. http://www.unicode.org/Public/UNIDATA/Blocks.txt */
+/* # Blocks-6.0.0.txt                                    */
+/* # Date: 2010-06-04, 11:12:00 PDT [KW]                 */
 static long ucs_range[]={
       0x0000, /* Basic Latin					0x00 */
       0x0080, /* Latin-1 Supplement				0x01 */
@@ -79,216 +81,228 @@
       0x0780, /* Thaana						0x0f */
       0x07C0, /* NKo						0x10 */
       0x0800, /* Samaritan					0x11 */
-      0x0900, /* Devanagari					0x12 */
-      0x0980, /* Bengali					0x13 */
-      0x0A00, /* Gurmukhi					0x14 */
-      0x0A80, /* Gujarati					0x15 */
-      0x0B00, /* Oriya						0x16 */
-      0x0B80, /* Tamil						0x17 */
-      0x0C00, /* Telugu						0x18 */
-      0x0C80, /* Kannada					0x19 */
-      0x0D00, /* Malayalam					0x1a */
-      0x0D80, /* Sinhala					0x1b */
-      0x0E00, /* Thai						0x1c */
-      0x0E80, /* Lao						0x1d */
-      0x0F00, /* Tibetan					0x1e */
-      0x1000, /* Myanmar					0x1f */
-      0x10A0, /* Georgian					0x20 */
-      0x1100, /* Hangul Jamo					0x21 */
-      0x1200, /* Ethiopic					0x22 */
-      0x1380, /* Ethiopic Supplement				0x23 */
-      0x13A0, /* Cherokee					0x24 */
-      0x1400, /* Unified Canadian Aboriginal Syllabics		0x25 */
-      0x1680, /* Ogham						0x26 */
-      0x16A0, /* Runic						0x27 */
-      0x1700, /* Tagalog					0x28 */
-      0x1720, /* Hanunoo					0x29 */
-      0x1740, /* Buhid						0x2a */
-      0x1760, /* Tagbanwa					0x2b */
-      0x1780, /* Khmer						0x2c */
-      0x1800, /* Mongolian					0x2d */
-      0x18B0, /* Unified Canadian Aboriginal Syllabics Extended	0x2e */
-      0x1900, /* Limbu						0x2f */
-      0x1950, /* Tai Le						0x30 */
-      0x1980, /* New Tai Lue					0x31 */
-      0x19E0, /* Khmer Symbols					0x32 */
-      0x1A00, /* Buginese					0x33 */
-      0x1A20, /* Tai Tham					0x34 */
-      0x1B00, /* Balinese					0x35 */
-      0x1B80, /* Sundanese					0x36 */
-      0x1C00, /* Lepcha						0x37 */
-      0x1C50, /* Ol Chiki					0x38 */
-      0x1CD0, /* Vedic Extensions				0x39 */
-      0x1D00, /* Phonetic Extensions				0x3a */
-      0x1D80, /* Phonetic Extensions Supplement			0x3b */
-      0x1DC0, /* Combining Diacritical Marks Supplement		0x3c */
-      0x1E00, /* Latin Extended Additional			0x3d */
-      0x1F00, /* Greek Extended					0x3e */
-      0x2000, /* General Punctuation				0x3f */
-      0x2070, /* Superscripts and Subscripts			0x40 */
-      0x20A0, /* Currency Symbols				0x41 */
-      0x20D0, /* Combining Diacritical Marks for Symbols	0x42 */
-      0x2100, /* Letterlike Symbols				0x43 */
-      0x2150, /* Number Forms					0x44 */
-      0x2190, /* Arrows						0x45 */
-      0x2200, /* Mathematical Operators				0x46 */
-      0x2300, /* Miscellaneous Technical			0x47 */
-      0x2400, /* Control Pictures				0x48 */
-      0x2440, /* Optical Character Recognition			0x49 */
-      0x2460, /* Enclosed Alphanumerics				0x4a */
-      0x2500, /* Box Drawing					0x4b */
-      0x2580, /* Block Elements					0x4c */
-      0x25A0, /* Geometric Shapes				0x4d */
-      0x2600, /* Miscellaneous Symbols				0x4e */
-      0x2700, /* Dingbats					0x4f */
-      0x27C0, /* Miscellaneous Mathematical Symbols-A		0x50 */
-      0x27F0, /* Supplemental Arrows-A				0x51 */
-      0x2800, /* Braille Patterns				0x52 */
-      0x2900, /* Supplemental Arrows-B				0x53 */
-      0x2980, /* Miscellaneous Mathematical Symbols-B		0x54 */
-      0x2A00, /* Supplemental Mathematical Operators		0x55 */
-      0x2B00, /* Miscellaneous Symbols and Arrows		0x56 */
-      0x2C00, /* Glagolitic					0x57 */
-      0x2C60, /* Latin Extended-C				0x58 */
-      0x2C80, /* Coptic						0x59 */
-      0x2D00, /* Georgian Supplement				0x5a */
-      0x2D30, /* Tifinagh					0x5b */
-      0x2D80, /* Ethiopic Extended				0x5c */
-      0x2DE0, /* Cyrillic Extended-A				0x5d */
-      0x2E00, /* Supplemental Punctuation			0x5e */
-      0x2E80, /* CJK Radicals Supplement			0x5f */
-      0x2F00, /* Kangxi Radicals				0x60 */
-      0x2FF0, /* Ideographic Description Characters		0x61 */
-      0x3000, /* CJK Symbols and Punctuation			0x62 */
-      0x3040, /* Hiragana					0x63 */
-      0x30A0, /* Katakana					0x64 */
-      0x3100, /* Bopomofo					0x65 */
-      0x3130, /* Hangul Compatibility Jamo			0x66 */
-      0x3190, /* Kanbun						0x67 */
-      0x31A0, /* Bopomofo Extended				0x68 */
-      0x31C0, /* CJK Strokes					0x69 */
-      0x31F0, /* Katakana Phonetic Extensions			0x6a */
-      0x3200, /* Enclosed CJK Letters and Months		0x6b */
-      0x3300, /* CJK Compatibility				0x6c */
-      0x3400, /* CJK Unified Ideographs Extension A		0x6d */
-      0x4DC0, /* Yijing Hexagram Symbols			0x6e */
-      0x4E00, /* CJK Unified Ideographs				0x6f */
-      0xA000, /* Yi Syllables					0x70 */
-      0xA490, /* Yi Radicals					0x71 */
-      0xA4D0, /* Lisu						0x72 */
-      0xA500, /* Vai						0x73 */
-      0xA640, /* Cyrillic Extended-B				0x74 */
-      0xA6A0, /* Bamum						0x75 */
-      0xA700, /* Modifier Tone Letters				0x76 */
-      0xA720, /* Latin Extended-D				0x77 */
-      0xA800, /* Syloti Nagri					0x78 */
-      0xA830, /* Common Indic Number Forms			0x79 */
-      0xA840, /* Phags-pa					0x7a */
-      0xA880, /* Saurashtra					0x7b */
-      0xA8E0, /* Devanagari Extended				0x7c */
-      0xA900, /* Kayah Li					0x7d */
-      0xA930, /* Rejang						0x7e */
-      0xA960, /* Hangul Jamo Extended-A				0x7f */
-      0xA980, /* Javanese					0x80 */
-      0xAA00, /* Cham						0x81 */
-      0xAA60, /* Myanmar Extended-A				0x82 */
-      0xAA80, /* Tai Viet					0x83 */
-      0xABC0, /* Meetei Mayek					0x84 */
-      0xAC00, /* Hangul Syllables				0x85 */
-      0xD7B0, /* Hangul Jamo Extended-B				0x86 */
-      0xD800, /* High Surrogates				0x87 */
-      0xDB80, /* High Private Use Surrogates			0x88 */
-      0xDC00, /* Low Surrogates					0x89 */
-      0xE000, /* Private Use Area				0x8a */
-      0xF900, /* CJK Compatibility Ideographs			0x8b */
-      0xFB00, /* Alphabetic Presentation Forms			0x8c */
-      0xFB50, /* Arabic Presentation Forms-A			0x8d */
-      0xFE00, /* Variation Selectors				0x8e */
-      0xFE10, /* Vertical Forms					0x8f */
-      0xFE20, /* Combining Half Marks				0x90 */
-      0xFE30, /* CJK Compatibility Forms			0x91 */
-      0xFE50, /* Small Form Variants				0x92 */
-      0xFE70, /* Arabic Presentation Forms-B			0x93 */
-      0xFF00, /* Halfwidth and Fullwidth Forms			0x94 */
-      0xFFF0, /* Specials                               	0x95 */
-      0x10000, /* Linear B Syllabary				0x96 */
-      0x10080, /* Linear B Ideograms				0x97 */
-      0x10100, /* Aegean Numbers				0x98 */
-      0x10140, /* Ancient Greek Numbers				0x99 */
-      0x10190, /* Ancient Symbols				0x9a */
-      0x101D0, /* Phaistos Disc					0x9b */
-      0x10280, /* Lycian					0x9c */
-      0x102A0, /* Carian					0x9d */
-      0x10300, /* Old Italic					0x9e */
-      0x10330, /* Gothic					0x9f */
-      0x10380, /* Ugaritic					0xa0 */
-      0x103A0, /* Old Persian					0xa1 */
-      0x10400, /* Deseret					0xa2 */
-      0x10450, /* Shavian					0xa3 */
-      0x10480, /* Osmanya					0xa4 */
-      0x10800, /* Cypriot Syllabary				0xa5 */
-      0x10840, /* Imperial Aramaic				0xa6 */
-      0x10900, /* Phoenician					0xa7 */
-      0x10920, /* Lydian					0xa8 */
-      0x10A00, /* Kharoshthi					0xa9 */
-      0x10A60, /* Old South Arabian				0xaa */
-      0x10B00, /* Avestan					0xab */
-      0x10B40, /* Inscriptional Parthian			0xac */
-      0x10B60, /* Inscriptional Pahlavi				0xad */
-      0x10C00, /* Old Turkic					0xae */
-      0x10E60, /* Rumi Numeral Symbols				0xaf */
-      0x11080, /* Kaithi					0xb0 */
-      0x12000, /* Cuneiform					0xb1 */
-      0x12400, /* Cuneiform Numbers and Punctuation		0xb2 */
-      0x13000, /* Egyptian Hieroglyphs				0xb3 */
-      0x1D000, /* Byzantine Musical Symbols			0xb4 */
-      0x1D100, /* Musical Symbols				0xb5 */
-      0x1D200, /* Ancient Greek Musical Notation		0xb6 */
-      0x1D300, /* Tai Xuan Jing Symbols				0xb7 */
-      0x1D360, /* Counting Rod Numerals				0xb8 */
-      0x1D400, /* Mathematical Alphanumeric Symbols		0xb9 */
-      0x1F000, /* Mahjong Tiles					0xba */
-      0x1F030, /* Domino Tiles					0xbb */
-      0x1F100, /* Enclosed Alphanumeric Supplement		0xbc */
-      0x1F200, /* Enclosed Ideographic Supplement		0xbd */
-      0x20000, /* CJK Unified Ideographs Extension B		0xbe */
-      0x2A700, /* CJK Unified Ideographs Extension C		0xbf */
-      0x2F800, /* CJK Compatibility Ideographs Supplement	0xc0 */
-      0x30000, /* reserved					0xc1 */
-      0x40000, /* reserved					0xc2 */
-      0x50000, /* reserved					0xc3 */
-      0x60000, /* reserved					0xc4 */
-      0x70000, /* reserved					0xc5 */
-      0x80000, /* reserved					0xc6 */
-      0x90000, /* reserved					0xc7 */
-      0xA0000, /* reserved					0xc8 */
-      0xB0000, /* reserved					0xc9 */
-      0xC0000, /* reserved					0xca */
-      0xD0000, /* reserved					0xcb */
-      0xE0000, /* Tags						0xcc */
-      0xE0100, /* Variation Selectors Supplement		0xcd */
-      0xF0000, /* Supplementary Private Use Area-A		0xce */
-      0x100000, /* Supplementary Private Use Area-B		0xcf */
+      0x0840, /* Mandaic					0x12 */
+      0x0900, /* Devanagari					0x13 */
+      0x0980, /* Bengali					0x14 */
+      0x0A00, /* Gurmukhi					0x15 */
+      0x0A80, /* Gujarati					0x16 */
+      0x0B00, /* Oriya						0x17 */
+      0x0B80, /* Tamil						0x18 */
+      0x0C00, /* Telugu						0x19 */
+      0x0C80, /* Kannada					0x1a */
+      0x0D00, /* Malayalam					0x1b */
+      0x0D80, /* Sinhala					0x1c */
+      0x0E00, /* Thai						0x1d */
+      0x0E80, /* Lao						0x1e */
+      0x0F00, /* Tibetan					0x1f */
+      0x1000, /* Myanmar					0x20 */
+      0x10A0, /* Georgian					0x21 */
+      0x1100, /* Hangul Jamo					0x22 */
+      0x1200, /* Ethiopic					0x23 */
+      0x1380, /* Ethiopic Supplement				0x24 */
+      0x13A0, /* Cherokee					0x25 */
+      0x1400, /* Unified Canadian Aboriginal Syllabics		0x26 */
+      0x1680, /* Ogham						0x27 */
+      0x16A0, /* Runic						0x28 */
+      0x1700, /* Tagalog					0x29 */
+      0x1720, /* Hanunoo					0x2a */
+      0x1740, /* Buhid						0x2b */
+      0x1760, /* Tagbanwa					0x2c */
+      0x1780, /* Khmer						0x2d */
+      0x1800, /* Mongolian					0x2e */
+      0x18B0, /* Unified Canadian Aboriginal Syllabics Extended	0x2f */
+      0x1900, /* Limbu						0x30 */
+      0x1950, /* Tai Le						0x31 */
+      0x1980, /* New Tai Lue					0x32 */
+      0x19E0, /* Khmer Symbols					0x33 */
+      0x1A00, /* Buginese					0x34 */
+      0x1A20, /* Tai Tham					0x35 */
+      0x1B00, /* Balinese					0x36 */
+      0x1B80, /* Sundanese					0x37 */
+      0x1BC0, /* Batak						0x38 */
+      0x1C00, /* Lepcha						0x39 */
+      0x1C50, /* Ol Chiki					0x3a */
+      0x1CD0, /* Vedic Extensions				0x3b */
+      0x1D00, /* Phonetic Extensions				0x3c */
+      0x1D80, /* Phonetic Extensions Supplement			0x3d */
+      0x1DC0, /* Combining Diacritical Marks Supplement		0x3e */
+      0x1E00, /* Latin Extended Additional			0x3f */
+      0x1F00, /* Greek Extended					0x40 */
+      0x2000, /* General Punctuation				0x41 */
+      0x2070, /* Superscripts and Subscripts			0x42 */
+      0x20A0, /* Currency Symbols				0x43 */
+      0x20D0, /* Combining Diacritical Marks for Symbols	0x44 */
+      0x2100, /* Letterlike Symbols				0x45 */
+      0x2150, /* Number Forms					0x46 */
+      0x2190, /* Arrows						0x47 */
+      0x2200, /* Mathematical Operators				0x48 */
+      0x2300, /* Miscellaneous Technical			0x49 */
+      0x2400, /* Control Pictures				0x4a */
+      0x2440, /* Optical Character Recognition			0x4b */
+      0x2460, /* Enclosed Alphanumerics				0x4c */
+      0x2500, /* Box Drawing					0x4d */
+      0x2580, /* Block Elements					0x4e */
+      0x25A0, /* Geometric Shapes				0x4f */
+      0x2600, /* Miscellaneous Symbols				0x50 */
+      0x2700, /* Dingbats					0x51 */
+      0x27C0, /* Miscellaneous Mathematical Symbols-A		0x52 */
+      0x27F0, /* Supplemental Arrows-A				0x53 */
+      0x2800, /* Braille Patterns				0x54 */
+      0x2900, /* Supplemental Arrows-B				0x55 */
+      0x2980, /* Miscellaneous Mathematical Symbols-B		0x56 */
+      0x2A00, /* Supplemental Mathematical Operators		0x57 */
+      0x2B00, /* Miscellaneous Symbols and Arrows		0x58 */
+      0x2C00, /* Glagolitic					0x59 */
+      0x2C60, /* Latin Extended-C				0x5a */
+      0x2C80, /* Coptic						0x5b */
+      0x2D00, /* Georgian Supplement				0x5c */
+      0x2D30, /* Tifinagh					0x5d */
+      0x2D80, /* Ethiopic Extended				0x5e */
+      0x2DE0, /* Cyrillic Extended-A				0x5f */
+      0x2E00, /* Supplemental Punctuation			0x60 */
+      0x2E80, /* CJK Radicals Supplement			0x61 */
+      0x2F00, /* Kangxi Radicals				0x62 */
+      0x2FF0, /* Ideographic Description Characters		0x63 */
+      0x3000, /* CJK Symbols and Punctuation			0x64 */
+      0x3040, /* Hiragana					0x65 */
+      0x30A0, /* Katakana					0x66 */
+      0x3100, /* Bopomofo					0x67 */
+      0x3130, /* Hangul Compatibility Jamo			0x68 */
+      0x3190, /* Kanbun						0x69 */
+      0x31A0, /* Bopomofo Extended				0x6a */
+      0x31C0, /* CJK Strokes					0x6b */
+      0x31F0, /* Katakana Phonetic Extensions			0x6c */
+      0x3200, /* Enclosed CJK Letters and Months		0x6d */
+      0x3300, /* CJK Compatibility				0x6e */
+      0x3400, /* CJK Unified Ideographs Extension A		0x6f */
+      0x4DC0, /* Yijing Hexagram Symbols			0x70 */
+      0x4E00, /* CJK Unified Ideographs				0x71 */
+      0xA000, /* Yi Syllables					0x72 */
+      0xA490, /* Yi Radicals					0x73 */
+      0xA4D0, /* Lisu						0x74 */
+      0xA500, /* Vai						0x75 */
+      0xA640, /* Cyrillic Extended-B				0x76 */
+      0xA6A0, /* Bamum						0x77 */
+      0xA700, /* Modifier Tone Letters				0x78 */
+      0xA720, /* Latin Extended-D				0x79 */
+      0xA800, /* Syloti Nagri					0x7a */
+      0xA830, /* Common Indic Number Forms			0x7b */
+      0xA840, /* Phags-pa					0x7c */
+      0xA880, /* Saurashtra					0x7d */
+      0xA8E0, /* Devanagari Extended				0x7e */
+      0xA900, /* Kayah Li					0x7f */
+      0xA930, /* Rejang						0x80 */
+      0xA960, /* Hangul Jamo Extended-A				0x81 */
+      0xA980, /* Javanese					0x82 */
+      0xAA00, /* Cham						0x83 */
+      0xAA60, /* Myanmar Extended-A				0x84 */
+      0xAA80, /* Tai Viet					0x85 */
+      0xAB00, /* Ethiopic Extended-A				0x86 */
+      0xABC0, /* Meetei Mayek					0x87 */
+      0xAC00, /* Hangul Syllables				0x88 */
+      0xD7B0, /* Hangul Jamo Extended-B				0x89 */
+      0xD800, /* High Surrogates				0x8a */
+      0xDB80, /* High Private Use Surrogates			0x8b */
+      0xDC00, /* Low Surrogates					0x8c */
+      0xE000, /* Private Use Area				0x8d */
+      0xF900, /* CJK Compatibility Ideographs			0x8e */
+      0xFB00, /* Alphabetic Presentation Forms			0x8f */
+      0xFB50, /* Arabic Presentation Forms-A			0x90 */
+      0xFE00, /* Variation Selectors				0x91 */
+      0xFE10, /* Vertical Forms					0x92 */
+      0xFE20, /* Combining Half Marks				0x93 */
+      0xFE30, /* CJK Compatibility Forms			0x94 */
+      0xFE50, /* Small Form Variants				0x95 */
+      0xFE70, /* Arabic Presentation Forms-B			0x96 */
+      0xFF00, /* Halfwidth and Fullwidth Forms			0x97 */
+      0xFFF0, /* Specials                               	0x98 */
+      0x10000, /* Linear B Syllabary				0x99 */
+      0x10080, /* Linear B Ideograms				0x9a */
+      0x10100, /* Aegean Numbers				0x9b */
+      0x10140, /* Ancient Greek Numbers				0x9c */
+      0x10190, /* Ancient Symbols				0x9d */
+      0x101D0, /* Phaistos Disc					0x9e */
+      0x10280, /* Lycian					0x9f */
+      0x102A0, /* Carian					0xa0 */
+      0x10300, /* Old Italic					0xa1 */
+      0x10330, /* Gothic					0xa2 */
+      0x10380, /* Ugaritic					0xa3 */
+      0x103A0, /* Old Persian					0xa4 */
+      0x10400, /* Deseret					0xa5 */
+      0x10450, /* Shavian					0xa6 */
+      0x10480, /* Osmanya					0xa7 */
+      0x10800, /* Cypriot Syllabary				0xa8 */
+      0x10840, /* Imperial Aramaic				0xa9 */
+      0x10900, /* Phoenician					0xaa */
+      0x10920, /* Lydian					0xab */
+      0x10A00, /* Kharoshthi					0xac */
+      0x10A60, /* Old South Arabian				0xad */
+      0x10B00, /* Avestan					0xae */
+      0x10B40, /* Inscriptional Parthian			0xaf */
+      0x10B60, /* Inscriptional Pahlavi				0xb0 */
+      0x10C00, /* Old Turkic					0xb1 */
+      0x10E60, /* Rumi Numeral Symbols				0xb2 */
+      0x11000, /* Brahmi					0xb3 */
+      0x11080, /* Kaithi					0xb4 */
+      0x12000, /* Cuneiform					0xb5 */
+      0x12400, /* Cuneiform Numbers and Punctuation		0xb6 */
+      0x13000, /* Egyptian Hieroglyphs				0xb7 */
+      0x16800, /* Bamum Supplement				0xb8 */
+      0x1B000, /* Kana Supplement				0xb9 */
+      0x1D000, /* Byzantine Musical Symbols			0xba */
+      0x1D100, /* Musical Symbols				0xbb */
+      0x1D200, /* Ancient Greek Musical Notation		0xbc */
+      0x1D300, /* Tai Xuan Jing Symbols				0xbd */
+      0x1D360, /* Counting Rod Numerals				0xbe */
+      0x1D400, /* Mathematical Alphanumeric Symbols		0xbf */
+      0x1F000, /* Mahjong Tiles					0xc0 */
+      0x1F030, /* Domino Tiles					0xc1 */
+      0x1F0A0, /* Playing Cards					0xc2 */
+      0x1F100, /* Enclosed Alphanumeric Supplement		0xc3 */
+      0x1F200, /* Enclosed Ideographic Supplement		0xc4 */
+      0x1F300, /* Miscellaneous Symbols And Pictographs		0xc5 */
+      0x1F600, /* Emoticons					0xc6 */
+      0x1F680, /* Transport And Map Symbols			0xc7 */
+      0x1F700, /* Alchemical Symbols				0xc8 */
+      0x20000, /* CJK Unified Ideographs Extension B		0xc9 */
+      0x2A700, /* CJK Unified Ideographs Extension C		0xca */
+      0x2B740, /* CJK Unified Ideographs Extension D		0xcb */
+      0x2F800, /* CJK Compatibility Ideographs Supplement	0xcc */
+      0x30000, /* reserved					0xcd */
+      0x40000, /* reserved					0xce */
+      0x50000, /* reserved					0xcf */
+      0x60000, /* reserved					0xd0 */
+      0x70000, /* reserved					0xd1 */
+      0x80000, /* reserved					0xd2 */
+      0x90000, /* reserved					0xd3 */
+      0xA0000, /* reserved					0xd4 */
+      0xB0000, /* reserved					0xd5 */
+      0xC0000, /* reserved					0xd6 */
+      0xD0000, /* reserved					0xd7 */
+      0xE0000, /* Tags						0xd8 */
+      0xE0100, /* Variation Selectors Supplement		0xd9 */
+      0xF0000, /* Supplementary Private Use Area-A		0xda */
+      0x100000, /* Supplementary Private Use Area-B		0xdb */
   /* Value over 0x10FFFF is illegal under Unicode,
      They are for some special use.  *** experimental ***  */
-      0x110000, /* Reserved					0xd0 */
-      0x120000, /* Reserved					0xd1 */
-      0x130000, /* Reserved					0xd2 */
-      0x140000, /* Reserved					0xd3 */
-      0x150000, /* Reserved					0xd4 */
-      0x160000, /* Reserved					0xd5 */
-      0x170000, /* Reserved					0xd6 */
-      0x180000, /* Reserved					0xd7 */
-      0x190000, /* Reserved					0xd8 */
-      0x1A0000, /* Reserved					0xd9 */
-      0x1B0000, /* Reserved					0xda */
-      0x1C0000, /* Reserved					0xdb */
-      0x1D0000, /* Reserved					0xdc */
-      0x1E0000, /* Reserved					0xdd */
-      0x1F0000, /* Reserved					0xde */
-      0x200000, /* Reserved					0xdf */
-      0x210000, /* Reserved					0xe0 */
-      0x220000, /* Reserved					0xe1 */
+      0x110000, /* Reserved					0xdc */
+      0x120000, /* Reserved					0xdd */
+      0x130000, /* Reserved					0xde */
+      0x140000, /* Reserved					0xdf */
+      0x150000, /* Reserved					0xe0 */
+      0x160000, /* Reserved					0xe1 */
+      0x170000, /* Reserved					0xe2 */
+      0x180000, /* Reserved					0xe3 */
+      0x190000, /* Reserved					0xe4 */
+      0x1A0000, /* Reserved					0xe5 */
+      0x1B0000, /* Reserved					0xe6 */
+      0x1C0000, /* Reserved					0xe7 */
+      0x1D0000, /* Reserved					0xe8 */
+      0x1E0000, /* Reserved					0xe9 */
+      0x1F0000, /* Reserved					0xea */
+      0x200000, /* Reserved					0xeb */
+      0x210000, /* Reserved					0xec */
+      0x220000, /* Reserved					0xed */
       CJK_CHAR_LIMIT
 };
 
diff -urN TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/upbibtex.ch uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/upbibtex.ch
--- TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/upbibtex.ch	2011-08-29 23:16:15 +0900
+++ uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/upbibtex.ch	2011-12-23 12:42:41 +0900
@@ -1,7 +1,7 @@
 @x
 @d banner=='This is pBibTeX, Version 0.99d-j0.33'
 @y
-@d banner=='This is upBibTeX, Version 0.99d-j0.33-u0.30'
+@d banner=='This is upBibTeX, Version 0.99d-j0.33-u1.00'
 @z
 
 @x
diff -urN TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/updvitype.ch uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/updvitype.ch
--- TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/updvitype.ch	2011-10-01 22:55:48 +0900
+++ uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/updvitype.ch	2011-12-23 12:40:49 +0900
@@ -1,7 +1,7 @@
 @x
 @d banner=='This is pDVItype, Version 3.6-p0.4'
 @y
-@d banner=='This is upDVItype, Version 3.6-p0.4-u0.30'
+@d banner=='This is upDVItype, Version 3.6-p0.4-u1.00'
 @z
 
 @x procedure initialize
diff -urN TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/uppltotf.ch uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/uppltotf.ch
--- TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/uppltotf.ch	2011-12-22 23:17:34 +0900
+++ uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/uppltotf.ch	2011-12-23 12:40:59 +0900
@@ -1,7 +1,7 @@
 @x
 @d banner=='This is pPLtoTF, Version 3.5-p1.8'
 @y
-@d banner=='This is upPLtoTF, Version 3.5-p1.8-u0.30'
+@d banner=='This is upPLtoTF, Version 3.5-p1.8-u1.00'
 @z
 
 @x
diff -urN TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/uptex-m.ch uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/uptex-m.ch
--- TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/uptex-m.ch	2011-12-22 23:17:34 +0900
+++ uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/uptex-m.ch	2012-01-14 12:16:05 +0900
@@ -1,4 +1,4 @@
-% This is a change file for upTeX u0.30
+% This is a change file for upTeX u1.00
 % By Takuji Tanaka.
 %
 % (02/26/2007) TTK  upTeX u0.01
@@ -29,13 +29,14 @@
 % (08/23/2009) TTK  upTeX u0.28
 % (01/31/2010) TTK  upTeX u0.29
 % (04/10/2010) TTK  upTeX u0.30
+% (01/15/2012) TTK  upTeX u1.00
 
 @x upTeX: banner
   {printed when p\TeX\ starts}
 @y
   {printed when p\TeX\ starts}
 @#
-@d upTeX_version_string=='-u0.30' {current up\TeX\ version}
+@d upTeX_version_string=='-u1.00' {current up\TeX\ version}
 @#
 @d upTeX_version==pTeX_version_string,upTeX_version_string
 @d upTeX_banner=='This is upTeX, Version 3.1415926',upTeX_version
@@ -216,23 +217,22 @@
 if (isinternalUPTEX) then begin
   { default: other_kchar }
   @t\hskip10pt@>kcat_code(@"0):=not_cjk;
-  @t\hskip10pt@>kcat_code(@"21):=hangul; { Hangul Jamo }
-  @+@t\1@>for k:=@"5F to @"61 do kcat_code(k):=kanji; { CJK Radicals Supplement .. Ideographic Description Characters }
-  @+@t\1@>for k:=@"63 to @"64 do kcat_code(k):=kana;  { Hiragana, Katakana }
-  @t\hskip10pt@>kcat_code(@"65):=kanji; { Bopomofo }
-  @t\hskip10pt@>kcat_code(@"66):=hangul; { Hangul Compatibility Jamo }
-  @+@t\1@>for k:=@"67 to @"69 do kcat_code(k):=kanji; { Kanbun .. CJK Strokes }
-  @t\hskip10pt@>kcat_code(@"6A):=kana; { Katakana Phonetic Extensions }
-  @t\hskip10pt@>kcat_code(@"6D):=kanji; { CJK Unified Ideographs Extension A }
-  @t\hskip10pt@>kcat_code(@"6F):=kanji; { CJK Unified Ideographs }
-  @t\hskip10pt@>kcat_code(@"7F):=hangul; { Hangul Jamo Extended-A }
-  @t\hskip10pt@>kcat_code(@"85):=hangul; { Hangul Syllables }
-  @t\hskip10pt@>kcat_code(@"86):=hangul; { Hangul Jamo Extended-B }
-  @t\hskip10pt@>kcat_code(@"8B):=kanji; { CJK Compatibility Ideographs }
-  { @t\hskip10pt@>kcat_code(@"94):=other_kchar; Halfwidth and Fullwidth Forms }
-  @t\hskip10pt@>kcat_code(@"BE):=kanji; { CJK Unified Ideographs Extension B }
-  @t\hskip10pt@>kcat_code(@"BF):=kanji; { CJK Unified Ideographs Extension C }
-  @t\hskip10pt@>kcat_code(@"C0):=kanji; { CJK Compatibility Ideographs Supplement }
+  @t\hskip10pt@>kcat_code(@"22):=hangul; { Hangul Jamo }
+  @+@t\1@>for k:=@"61 to @"63 do kcat_code(k):=kanji; { CJK Radicals Supplement .. Ideographic Description Characters }
+  @+@t\1@>for k:=@"65 to @"66 do kcat_code(k):=kana;  { Hiragana, Katakana }
+  @t\hskip10pt@>kcat_code(@"67):=kanji; { Bopomofo }
+  @t\hskip10pt@>kcat_code(@"68):=hangul; { Hangul Compatibility Jamo }
+  @+@t\1@>for k:=@"69 to @"6B do kcat_code(k):=kanji; { Kanbun .. CJK Strokes }
+  @t\hskip10pt@>kcat_code(@"6C):=kana; { Katakana Phonetic Extensions }
+  @t\hskip10pt@>kcat_code(@"6F):=kanji; { CJK Unified Ideographs Extension A }
+  @t\hskip10pt@>kcat_code(@"71):=kanji; { CJK Unified Ideographs }
+  @t\hskip10pt@>kcat_code(@"81):=hangul; { Hangul Jamo Extended-A }
+  @t\hskip10pt@>kcat_code(@"88):=hangul; { Hangul Syllables }
+  @t\hskip10pt@>kcat_code(@"89):=hangul; { Hangul Jamo Extended-B }
+  @t\hskip10pt@>kcat_code(@"8E):=kanji; { CJK Compatibility Ideographs }
+  { @t\hskip10pt@>kcat_code(@"97):=other_kchar; Halfwidth and Fullwidth Forms }
+  @t\hskip10pt@>kcat_code(@"B9):=kana; { Kana Supplement }
+  @+@t\1@>for k:=@"C9 to @"CC do kcat_code(k):=kanji; { CJK Unified Ideographs Extension B .. CJK Compatibility Ideographs Supplement }
   @t\hskip10pt@>kcat_code(@"FE):=kana; { Fullwidth digit and latin alphabet }
   @t\hskip10pt@>kcat_code(@"FF):=kana; { Halfwidth katakana }
 end else begin
diff -urN TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/uptftopl.ch uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/uptftopl.ch
--- TeXLive_svn_r25014/Build/source/texk/web2c/uptexdir/uptftopl.ch	2011-12-22 23:17:34 +0900
+++ uptex-1.00-20120115a/Build/source/texk/web2c/uptexdir/uptftopl.ch	2011-12-23 12:41:09 +0900
@@ -1,7 +1,7 @@
 @x
 @d banner=='This is pTFtoPL, Version 3.2-p1.7'
 @y
-@d banner=='This is upTFtoPL, Version 3.2-p1.7-u0.30'
+@d banner=='This is upTFtoPL, Version 3.2-p1.7-u1.00'
 @z
 
 @x
